Handle non-zero length nt_array_ptrs

mgrang opened the pull request:
We can do bounds-widening for nt_array_ptrs variables whose upper bound is not
just the variables itself (that is, non-zero length array_ptrs).  Consider,
nt_array_ptr p : bounds(e1, e2); where e1 and e2 are expressions and e2 is of
the form (e + k) To widen the bounds, we look for derefs of the form "*(e2 +
i)".

Example:
nt_array_ptr<T> p : bounds(p, p + i);

if (*(p + i)) // widen by 1
  if (*(p + i + 1)) // widen by 2
    if (*(p + i + 2)) // widen by 3

if (*(p + i)) // widen by 1
  if (*(p + i + 2)) // no widening

if (*p) // no widening
  if (*(p+ 1)) // no widening
    if (*(p + 2)) // no widening

mgrang commented:
This PR is a replacement for https://github.com/microsoft/checkedc-clang/pull/746.

mgrang commented on clang/lib/Sema/BoundsAnalysis.cpp:
```diff
@@ -136,105 +136,278 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
-  // For conditions of the form "if (*p)".
-  if (IsDeclOperand(E)) {
-    const DeclRefExpr *D = GetDeclOperand(E);
+  const auto *RBE = dyn_cast<RangeBoundsExpr>(BE);
```
By this time, non-range bounds expressions like "count(e1)" are already canonicalized to "bounds(p, p + e1)". And it is trivial to check for bounds(unknown) and bounds(any) and return. Are there other forms of bounds expressions that I may be missing?

mgrang commented on clang/lib/Sema/BoundsAnalysis.cpp:
```diff
@@ -136,105 +136,278 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
+        return V;
+  }
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
```
I don't think so. I am already checking for the binary operator inside SplitIntoBaseOffset and if the operator is not BO_Add then we do not even reach here.

mgrang commented:
@dtarditi I have updated the PR with logic for widening all ntptrs in scope based on matching their upper bounds with the deref base. Could you please review this?

mgrang commented on clang/lib/Sema/BoundsAnalysis.cpp:
```diff
@@ -136,105 +136,344 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
+    //   if (*(p + i + 1)) // Offset = 1 < UBO ==> no widening
+    //     if (*(p + i + 2)) // Offset = 2 == UBO ==> widen by 1
+    //       if (*(p + i + 3)) // Offset = 3 > UBO ==> widen by 2
+    if (Lex.CompareAPInt(DerefOffsetVal, UpperOffsetVal) ==
```
I am handling this while computing the union. We only widen p if the bounds of p in (In - Kill) == Gen[p]. So we would not widen the bounds by 4 in case of `if (*(p + i + 5)`.

mgrang replied:
I have added a test where widening should not happen because of a missing test of an element.

dtarditi commented on clang/lib/Sema/BoundsAnalysis.cpp:
```diff
@@ -136,105 +136,344 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
+    const llvm::APInt UpperOffsetVal = GetAPIntVal(UpperOffset);
+
+    // We cannot widen the bounds if the offset in the deref expr is less than
+    // the offset in the upper bounds expr. For example:
```
Yes, you are correct in your reply to my comments.  Could you modify the comment in line 334 to say "the offset in the declared upper bounds expr.  For example.?

dtarditi commented on clang/lib/Sema/BoundsAnalysis.cpp:
```diff
@@ -136,105 +136,344 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
+        Lexicographic::Result::LessThan)
+      continue;
+
+    // (DerefOffset - UpperOffset) gives us the amount by which the ntptr should
```
Could you reword this?   The phase "should be widened" confused me.  Perhaps "(DerefOffset - UpperOffset) gives the  offset of the memory dereference relative to the declared upper bound expression.  This offset is used in the widening computation in ComputeOutSets."

dtarditi commented on clang/include/clang/Sema/BoundsAnalysis.h:
```diff
@@ -253,6 +254,40 @@ namespace clang {
+    // Get the APInt value from an expression representing an IntegerLiteral.
+    // @param[in] E represents an IntegerLiteral.
+    // @return Return the APInt value for E. Return APInt(0) if E is nullptr.
+    const llvm::APInt GetAPIntVal(const Expr *E) const;
```
I think it would be better to use the function Expr::isIntegerConstantExpr instead of defining your own functions for recognizing constant expressions.  It does return an llvm::APSInt,  which I believe is what you want.  A clang IntegerLiteral expression uses an APInt (unsigned integer) to represent the bit pattern for the integer defined but the literal, but the bit pattern needs to be interpreted based on whether the *type* of the literal is a signed or unsigned integer type.   The function splitIntoBaseAndOffset in SemaBounds.cpp shows how to use isIntegerConstantExpr.

mgrang commented:
@dtarditi  Thanks for your review! I have addressed your comments in my latest update. Here are the updates:

- Collected all nt_array_ptr variables defined in the current function.
- Replaced custom IsIntegerOperand and GetAPIntVal functions with isIntegerConstantExpr.
- Made the order of printing of VarDecls deterministic.
- Updated comments.
